src/coreclr/jit/optcse.cpp

```diff
@@ -500,6 +500,9 @@ unsigned Compiler::optValnumCSE_Index(GenTree* tree, Statement* stmt)
             // we don't allow/expect it to be set when we need a reloc
             //
             assert((key & TARGET_SIGN_BIT) == 0);
+
+            // Don't use the CoercedConstantValue value as the key
+            key = vnLibNorm;
```

Comment:
Seems like this would be less confusing if you didn't set and reset?
```C++
        assert(vnStore->IsVNConstant(vnLibNorm));

        // We don't share small offset constants when we require a reloc
        //
        if (!tree->AsIntConCommon()->ImmedValNeedsReloc(this))
        {
            // Make constants that have the same upper bits use the same key
            key           = vnStore->CoercedConstantValue<size_t>(vnLibNorm);
            key           = Encode_Shared_Const_CSE_Value(key);
            isSharedConst = true;
        }
        else
        {
            key = vnLibNorm;
```

Reply:
Fair enough, I will make this change.

Reply:
@AndyAyersMS

I restructured this slightly and added comments that make it clear that the key encodes a shared constant only when isSharedConst is true.  I also moved this assert and changed it to cover both cases:

`assert(isSharedConst == Is_Shared_Const_CSE(key));`